internal/characters/qiqi/qiqi.go

```diff
@@ -0,0 +1,176 @@
+package qiqi
+
+import (
+	"github.com/genshinsim/gcsim/pkg/character"
+	"github.com/genshinsim/gcsim/pkg/core"
+)
+
+func init() {
+	core.RegisterCharFunc("qiqi", NewChar)
+}
+
+type char struct {
+	*character.Tmpl
+	talismanExpiry    []int
+	talismanICDExpiry []int
+	c4ICDExpiry       int
+	skillLastUsed     int
+}
+
+// TODO: Not implemented - C6 (revival mechanic, not suitable for sim)
+// C4 - Enemy Atk reduction, not useful in this sim version
+func NewChar(s *core.Core, p core.CharacterProfile) (core.Character, error) {
+	c := char{}
+	t, err := character.NewTemplateChar(s, p)
+	if err != nil {
+		return nil, err
+	}
+	c.Tmpl = t
+	c.Energy = 80
+	c.EnergyMax = 80
+	c.Weapon.Class = core.WeaponClassSword
+	c.NormalHitNum = 5
+	c.BurstCon = 3
+	c.SkillCon = 5
+
+	c.talismanExpiry = make([]int, len(c.Core.Targets))
+	c.talismanICDExpiry = make([]int, len(c.Core.Targets))
+	c.skillLastUsed = 0
+
+	c.talismanHealHook()
+	c.onNACAHitHook()
+	c.a1()
+
+	return &c, nil
+}
+
+func (c *char) talismanHealHook() {
+	c.Core.Events.Subscribe(core.OnDamage, func(args ...interface{}) bool {
+		t := args[0].(core.Target)
+		ds := args[1].(*core.Snapshot)
+
+		if c.talismanExpiry[t.Index()] < c.Core.F {
+			return false
+		}
+		if c.talismanICDExpiry[t.Index()] >= c.Core.F {
+			return false
+		}
+
+		// Talisman healing is dynamic, so need to create a new snapshot instance for it
+		dsHeal := c.Snapshot(
+			"Qiqi Talisman Healing",
+			core.AttackTagElementalBurst,
+			core.ICDTagNone,
+			core.ICDGroupDefault,
+			core.StrikeTypeDefault,
+			core.Cryo,
+			0,
+			0,
+		)
+		healAmt := c.heal(&dsHeal, burstHealPer, burstHealFlat, c.TalentLvlBurst())
+		c.Core.Health.HealIndex(c.Index, ds.ActorIndex, healAmt)
+		c.talismanICDExpiry[t.Index()] = c.Core.F + 60
+
+		c.Core.Log.Debugw("Qiqi Talisman Healing", "frame", c.Core.F, "event", core.LogCharacterEvent, "char", c.Index, "target", t.Index(), "healed_char", ds.ActorIndex, "talisman_expiry", c.talismanExpiry[t.Index()], "talisman_healing_icd", c.talismanICDExpiry[t.Index()], "healed_amt", healAmt)
+
+		return false
+	}, "talisman-heal-hook")
+}
+
+// Handles C2, A4, and skill NA/CA on hit hooks
+// Additionally handles burst Talisman hook - can't be done another way since Talisman is applied before the burst damage is dealt
```

Comment:
rather than queueing damage in burst, why not just add a task that takes a snapshot, applies talisman, then call ApplyDamage? wouldn't that achieve the goal of applying talisman first before triggering damage?

Reply:
Talisman is supposed to be applied on hit. I'm not a fan of manually applying it in the burst since that's more liable to break depending on how things interact in the future.